File: android/app/src/main/java/in/stewardmd/app/AppOrientationPlugin.java
package in.stewardmd.app;

import android.app.Activity;
import android.content.pm.ActivityInfo;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/**
 * Screen-orientation control for StewardMD. The app is portrait-locked app-wide
 * (android:screenOrientation="portrait" on MainActivity); the antibiogram grid calls
 * unlock() so it can be read in landscape, then lockPortrait() again on close.
 *
 * SCREEN_ORIENTATION_FULL_SENSOR honours the device's physical orientation but still
 * respects the system auto-rotate setting — if the user has auto-rotate off, the screen
 * won't physically turn even after unlock (expected OS behaviour).
 */
@CapacitorPlugin(name = "AppOrientation")
public class AppOrientationPlugin extends Plugin {

    @PluginMethod
    public void lockPortrait(PluginCall call) {
        setOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        call.resolve();
    }

    @PluginMethod
    public void unlock(PluginCall call) {
        setOrientation(ActivityInfo.SCREEN_ORIENTATION_FULL_SENSOR);
        call.resolve();
    }

    private void setOrientation(final int orientation) {
        final Activity activity = getActivity();
        if (activity == null) return;
        activity.runOnUiThread(() -> activity.setRequestedOrientation(orientation));
    }
}

File: local-plugins/capacitor-app-orientation/ios/Sources/AppOrientationPlugin/AppOrientationPlugin.swift
import Foundation
import Capacitor
import UIKit

/**
 * Screen-orientation control for StewardMD (iOS).
 *
 * The app is portrait-locked app-wide; the antibiogram grid calls unlock() so it can be
 * read in landscape, then lockPortrait() again on close. Exposed to JS as
 * `Capacitor.Plugins.AppOrientation.lockPortrait()` / `.unlock()`.
 *
 * The plugin does not own the UIKit state — it posts a notification and AppDelegate (which
 * implements `application(_:supportedInterfaceOrientationsFor:)`) updates the mask and
 * requests the geometry change. This keeps the plugin package decoupled from the app target.
 */
@objc(AppOrientationPlugin)
public class AppOrientationPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AppOrientationPlugin"
    public let jsName = "AppOrientation"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "lockPortrait", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unlock", returnType: CAPPluginReturnPromise)
    ]

    // Shared with AppDelegate via a string key so no cross-module symbol dependency is needed.
    public static let notificationName = Notification.Name("SMDSetOrientation")

    @objc func lockPortrait(_ call: CAPPluginCall) {
        post(UIInterfaceOrientationMask.portrait.rawValue)
        call.resolve()
    }

    @objc func unlock(_ call: CAPPluginCall) {
        post(UIInterfaceOrientationMask.allButUpsideDown.rawValue)
        call.resolve()
    }

    private func post(_ mask: UInt) {
        DispatchQueue.main.async {
            NotificationCenter.default.post(
                name: AppOrientationPlugin.notificationName,
                object: nil,
                userInfo: ["mask": mask]
            )
        }
    }
}
